### PR TITLE
RockPaperScissors and Parser

### PR DESCRIPTION
**Pull request by kakaly**
**RockPaperScissors**

* game.py has the Game base class.
* rock_paper_scissors.py has the RockPaperScissors class which is a child class of Game class. We are overriding the decide_winner() method of base class to implement the logic in the child class
* rps.py has the instantiation of RockPaperScissors
* tests.py has the tests for Game and RockPaperScissors's methods
* mock_true.py mocks the infinite while loop "True" boolean

**Console output**

```
DN525v8s:codingfun karthik$ python3 game/rps.py 

Game: Rock, Paper, Scissors


Select 'r' for Rock, 'p' for Paper, 's' for Scissors, 'e' for exit
r
Computer played: p
You lost!!

Select 'r' for Rock, 'p' for Paper, 's' for Scissors, 'e' for exit
p
Computer played: r
You won!!

Select 'r' for Rock, 'p' for Paper, 's' for Scissors, 'e' for exit
q
Invalid input: q, Valid inputs: ['r', 'p', 's']

Select 'r' for Rock, 'p' for Paper, 's' for Scissors, 'e' for exit
e
DN525v8s:codingfun karthik$ 
```
```
DN525v8s:codingfun karthik$ pytest game/tests.py 
============================================================================================================ test session starts =============================================================================================================
platform darwin -- Python 2.7.13, pytest-3.5.1, py-1.5.3, pluggy-0.6.0
rootdir: /Users/karthik/codingfun, inifile:
collected 8 items                                                                                                                                                                                                                            

game/tests.py ...x....                                                                                                                                                                                                                 [100%]

==================================================================================================== 7 passed, 1 xfailed in 0.08 seconds =====================================================================================================
DN525v8s:codingfun karthik$ 
```

**Parser**

* get_stats.py has the ParseFile class and its related methods 

**Console output**

```
DN525v8s:codingfun karthik$ python3 parser/get_stats.py parser/sample_log.txt 

Old:
 (Time, Min): ('22:55:19.444', 1.463), (Time, Mid): ('23:05:40.224', 17.73), (Time, Max): ('23:19:52.161', 107.5), Average: 25.925781249999996

New:
 (Time, Min): ('22:55:19.444', 2.12), (Time, Mid): ('23:17:37.920', 25.37), (Time, Max): ('23:19:52.161', 65.44), Average: 26.536500000000004

DN525v8s:codingfun karthik$ 
```

### CODE REVIEW

**Comment by tonywtao** [comment]
Thanks for submitting such a good pull request for reviewing.

File: game/game.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+    def play_game(self):
+        """Logic for playing the game.
+
+        Gets input from the user, validates it and
+        calculates the winner and prints it.
+
+        """
+        while True:
+            user_input = input(self.instruction)
```
Comment (tonywtao) [inline game/game.py]: It would be nice if you can make user input case insensitive.

File: game/rock_paper_scissors.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    def decide_winner(self, user_input, computer_input):
+        """Decides the winner based on the user input and
+        the computer's input
+
+        Rock blunts scissors, Paper covers rock,
+        Scissors cut paper
+
+        """
+        if user_input == 'p' and computer_input == 'r':
```
Comment (tonywtao) [inline game/rock_paper_scissors.py]: If we can define a user winning rule map in advance, the code to decide winner can be simplified a bit more.

File: parser/get_stats.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        """
+        result = min(self.metrics.items(), key=lambda x: x[1][1])
+        return (result[0].split("_")[0], result[1][1])
+
+    def get_mid_new(self):
+        """Gets the mid new value and returns
+        (time, mid new value)
+
+        """
+        mid_index = len(self.metrics)/2 if len(self.metrics)%2 == 0 else len(self.metrics)/2 + 1
```
Comment (tonywtao) [inline parser/get_stats.py]: (len(self.metrics) + 1)/2 will work.

File: parser/get_stats.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            line = fp.readline()
+            while line:
+                self.__parse_line(line)
+                line = fp.readline()
+        if self.metrics:
+            self.print_stats()
+        else:
+            raise Exception("No stats found in the file")
+
+    def __parse_line(self, line):
```
Comment (tonywtao) [inline parser/get_stats.py]: leading a method name with double underline seems unneccessary.

File: parser/get_stats.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        """Parses a line and adds the old value,
+        new value and time into the dict in the form,
+        {time_old_new: [old, new]}
+
+        """
+        line.strip()
+        if OLD in line and NEW in line:
+            old = line[line.index(OLD) + len(OLD):line.index(OLD) + len(OLD) + 5]
+            new = line[line.index(NEW) + len(NEW):line.index(NEW) + len(NEW) + 5]
```
Comment (tonywtao) [inline parser/get_stats.py]: Trimming old/new rate value using fixed length can result in data loss, as the rate value can vary from 0.001 to 999.999. It'll be better to use regular express I think.

File: parser/get_stats.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+    def get_max_new(self):
+        """Gets the max new value and returns
+        (time, max new value)
+
+        """
+        result = max(self.metrics.items(), key=lambda x: x[1][1])
+        return (result[0].split("_")[0], result[1][1])
+
+    def get_avg_new(self):
```
Comment (tonywtao) [inline parser/get_stats.py]: To define get_avg(new or old) will result in less duplicated code. Same for get_max/min/mid_xxx.

File: game/game.py (hunk 1 above)
Comment (kakaly) [reply game/game.py]: Thanks for the feedback. Yes, I agree with that too.

File: game/rock_paper_scissors.py (hunk 2 above)
Comment (kakaly) [reply game/rock_paper_scissors.py]: I agree. That will make it simpler and easier to read.

File: parser/get_stats.py (hunk 5 above)
Comment (kakaly) [reply parser/get_stats.py]: Oh okay. I wasn't entirely sure about the data loss part. Agreed, regex will make it more robust.

**Comment by kakaly** [comment]
Thanks @tonywtao for reviewing and giving me a feedback.

**Pull request closed without merging**